### ISSUE TITLE
Optimization: Fold redundant input parameters loads

### ISSUE THREAD

**Issue by tgiphil** [mosa/MOSA-Project#81]
A set of virtual registers that are defined only once and loaded from same method input parameter on the stack and that input parameter is never updated, can be folded into a single load.

Example:

In [System.String::IndexOfAny(System.Char[])](https://github.com/mosa/MOSA-Project/blob/master/Source/Korlib/System/String.cs#L458)

```
L_0020 X86.Mov V_6 [System.String] <= this [EBP+4h] [System.String]
L_0021 X86.Mov V_7 [System.Char[]] <= anyOf [EBP+4h] [System.Char[]]
L_0023 X86.Mov V_9 [System.String] <= this [EBP+4h] [System.String]
```

In this example, the _this_ variable should only be loaded at once into a single virtual register. 

The initial load must be placed before all uses from all paths. 

A native implementation can place the load in the first block during an early IR stage. This could be helpful for some SSA optimizations.

**Comment by tgiphil** [comment]
Plan: Insert load parameter instruction into the first block that dominates all other blocks which use the parameter. The load will move the parameter to a virtual instruction. The load instruction is placed at the end of the block, or if the block contains any existing load instruction, then immediately prior to that instruction. Replace all other load instructions with a move from the virtual instruction.

**Comment by tgiphil** [comment]
Alternative longer-term option: Similar to before, but instead move all the load instructions to the beginning of the first block. And write another optimization stage that re-orders instructions to reduce the distance between assignments (results) and uses. The register allocator will then "best-fit" the live ranges.

**Comment by tgiphil** [comment]
The value numbering optimization implemented this optimization and removes redundant parameter loads.
